### PR TITLE
Stable cache keys using DB session ids in SSTs

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: Use DB session ids in SST table properties to make cache keys
stable across DB re-open and copy / move / restore / etc.

These new cache keys are currently only enabled when FileSystem does not
provide GetUniqueId. For now, they are typically larger, so slightly
less efficient.

Relevant to #7405

This change has a minor regression in PersistentCache functionality:
metaindex blocks are no longer cached in PersistentCache. Table properties
blocks already were not but ideally should be. I didn't spent effort to
fix & test these issues because we don't believe PersistentCache is used much
if at all and expect SecondaryCache to replace it. (Though PRs are welcome.)

FIXME: there is more to be fixed for stable cache keys on external SST files

Test Plan: new unit test added, which fails when disabling new
functionality

### CODE REVIEW

File: table/block_based/block_based_table_reader.cc [hunk 1]
```diff
@@ -370,27 +371,36 @@ Cache::Handle* BlockBasedTable::GetEntryFromCache(
...
+        file_num);
   }
   if (rep->table_options.block_cache_compressed != nullptr) {
     GenerateCachePrefix<Cache, FSRandomAccessFile>(
         rep->table_options.block_cache_compressed.get(), rep->file->file(),
         &rep->compressed_cache_key_prefix[0],
-        &rep->compressed_cache_key_prefix_size, "", cur_file_num);
+        &rep->compressed_cache_key_prefix_size, db_session_id, file_num);
+  }
+  if (rep->table_options.persistent_cache != nullptr) {
```
Comment (zhichao-cao) [inline table/block_based/block_based_table_reader.cc:390]: I guess, no one is using the persistent cache. Maybe we can deperate it in the future when secondary cache is ready

Comment (pdillinger) [reply table/block_based/block_based_table_reader.cc:390]: Microsoft might be using it. See #6932 @zhli1142015

File: table/block_based/block_based_table_reader.cc [hunk 2]
```diff
@@ -619,6 +626,33 @@ Status BlockBasedTable::Open(
...
+      if (file_num > 0) {
+        db_session_id = rep->table_properties->db_session_id;
+      }
+    } else {
+      // External (ingested) SST file - should not use current file number
+      // (changed from original), so that same file ingested into different DBs
+      // can share block cache entries. Although they can modify the embedded
+      // global_seqno, that information is not currently cached under these
+      // portable/stable keys.
+      // Note: For now, each external SST file gets its own unique session id.
```
Comment (zhichao-cao) [inline table/block_based/block_based_table_reader.cc:649]: Maybe also mention that cc->NewId() will be used instead file_number as the suffix.

**Review by zhichao-cao — APPROVED** [review]
LGTM, thanks for the PR, it will make the db_session_id stable when db is reopened.

File: table/block_based/block_based_table_reader.cc (hunk 2 above)
Comment (pdillinger) [reply table/block_based/block_based_table_reader.cc:649]: Oops! My intention was to use a fixed file number as suffix but 0 means "unknown."

I will fix this and add to unit test.

Comment (pdillinger) [reply table/block_based/block_based_table_reader.cc:649]: Actually, fixing it requires a bit more intervention than I want to throw into this PR after approval.

Comment (zhichao-cao) [reply table/block_based/block_based_table_reader.cc]: I will take a look after you push the updates.

**Pull request closed without merging**